src/Compiler/Checking/CheckDeclarations.fs

```diff
@@ -563,26 +563,38 @@ module TcRecdUnionAndEnumDeclarations =
         let unionCasesR = unionCases |> List.map (TcUnionCaseDecl cenv env parent thisTy thisTyInst tpenv hasRQAAttribute) 
         unionCasesR |> CheckDuplicates (fun uc -> uc.Id) "union case" 
 
-    let TcEnumDecl cenv env parent thisTy fieldTy (SynEnumCase(attributes=Attributes synAttrs; ident= SynIdent(id,_); value=v; xmlDoc=xmldoc; range=m)) =
+    let TcEnumCaseConstDecl cenv env parent attrs thisTy caseRange (caseIdent: Ident) (xmldoc: PreXmlDoc) value =
+        let vis, _ = ComputeAccessAndCompPath env None caseRange None None parent
+        let vis = CombineReprAccess parent vis
+        if caseIdent.idText = "value__" then errorR(Error(FSComp.SR.tcNotValidEnumCaseName(), caseIdent.idRange))
+        let checkXmlDocs = cenv.diagnosticOptions.CheckXmlDocs
+        let xmlDoc = xmldoc.ToXmlDoc(checkXmlDocs, Some [])
+        Construct.NewRecdField true (Some value) caseIdent false thisTy false false [] attrs xmlDoc vis false
+
+    let TcEnumDecl cenv env tpenv parent thisTy fieldTy (SynEnumCase(attributes=Attributes synAttrs; ident= SynIdent(id,_); value=v; xmlDoc=xmldoc; range=m)) =
         let attrs = TcAttributes cenv env AttributeTargets.Field synAttrs
-        
+
         match v with 
-        | SynConst.Bytes _
-        | SynConst.UInt16s _
-        | SynConst.UserNum _ -> error(Error(FSComp.SR.tcInvalidEnumerationLiteral(), m))
-        | _ -> 
+        | SynExpr.Const (constant = SynConst.Bytes _ | SynConst.UInt16s _ | SynConst.UserNum _) ->
+            error(Error(FSComp.SR.tcInvalidEnumerationLiteral(), m))
+        | SynExpr.Const (v, _) -> 
             let v = TcConst cenv fieldTy m env v
-            let vis, _ = ComputeAccessAndCompPath env None m None None parent
-            let vis = CombineReprAccess parent vis
-            if id.idText = "value__" then errorR(Error(FSComp.SR.tcNotValidEnumCaseName(), id.idRange))
-            let checkXmlDocs = cenv.diagnosticOptions.CheckXmlDocs
-            let xmlDoc = xmldoc.ToXmlDoc(checkXmlDocs, Some [])
-            Construct.NewRecdField true (Some v) id false thisTy false false [] attrs xmlDoc vis false
-      
-    let TcEnumDecls (cenv: cenv) env parent thisTy enumCases =
+            TcEnumCaseConstDecl cenv env parent attrs thisTy m id xmldoc v
+        | _ ->
+            if cenv.g.langVersion.SupportsFeature LanguageFeature.ArithmeticInLiterals then
+                let expr, actualTy, _ = TcExprOfUnknownType cenv env tpenv v
+                UnifyTypes cenv env m fieldTy actualTy
```

Comment:
Bit of a sidenote, the signature `val UnifyTypes: cenv: TcFileState -> env: TcEnv -> m: range -> actualTy: TType -> expectedTy: TType -> unit` might be incorrect - the last parameter seems to be used as the actual type.

Reply:
I'm confused by what constitutes an expected and actual type.

Checking `if` expressions - unit is the expected `then` branch type when there is no `else`

```fsharp
if not isRecovery && Option.isNone synElseExprOpt then
     UnifyTypes cenv env m g.unit_ty overallTy.Commit
```

Checking `do` expressions - unit is the expected time.

```fsharp
| SynExpr.Do (synInnerExpr, m) ->
    UnifyTypes cenv env m overallTy.Commit g.unit_ty
```

Reply:
The g.unity_ty is the expected one, that matches also my understanding of expected vs. actual. 
The usages seem to confirm this view, with only a few outliers => BUT, changing things around now can lead to different (inverted) error message. So it should be first checked if the changed meaning is better.

![image](https://user-images.githubusercontent.com/46543583/210974890-d22ddcb9-3477-45ed-99d2-bc766650a235.png)